Oracle.md

```diff
@@ -0,0 +1,190 @@
+# Oracle specifications
+
+## Introduction
+
+For the purpose of these specifications, an event is a digital representation of a real world fact.
+An [oracle](./Introduction.md#Oracle) is an entity that commits to publishing a signature over a (number of) event outcome(s) ahead of time by releasing one or more [R-values](./Introduction.md#R-value) as well as necessary information for two parties to build a set of [CETs](./Introduction.md#Contract-Execution-Transaction-(CET)).
+This necessary information is committed in a so-called [_event descriptor_](#Event-descriptor) that will be further detailed in this document.
+Descriptors are classified and discovered through a namespace presently represented as Unique Resource Identifier (URI) paths.
+
+## Table of Contents
+
+- [Event descriptor](#event-descriptor)
+    - [Simple enumeration](#simple-enumeration)
+        - [Example: Weather tomorrow](#example-weather-tomorrow)
+    - [Range](#range)
+        - [Example: tomorrow's temperature](#example-tomorrows-temperature)
+    - [Large range](#large-range)
+        - [Example: BTC/USD rate](#example-btcusd-rate)
+    - [Serialization of event descriptors](#serialization-of-event-descriptors)
+        - [Version 0 `external_event_descriptor`](#version-0-external_event_descriptor)
+        - [Version 0 `enum_event_descriptor`](#version-0-enum_event_descriptor)
+        - [Version 0 `range_event_descriptor`](#version-0-range_event_descriptor)
+        - [Version 0 `large_range_event_descriptor`](#version-0-large_range_event_descriptor)
+    - [Oracle events](#oracle-events)
+        - [Version 0 `oracle_event`](#version-0-oracle_event)
+- [Oracle announcements](#oracle-announcements)
+        - [Version 0 `oracle_announcement`](#version-0-oracle_announcement)
+
+
+## Event descriptor
+
+An event descriptor provides information to clients about an event for which an oracle plans on releasing a signature over its outcome.
+The provided information should be sufficient for a client to create a set of CETs that will cover all possible outcomes of the event.
+
+Here we assume that an event outcome can be represented either as a set of strings or numbers.
+Four kinds of event outcomes are defined (note that a single event can be composed of several types).
+
+### Simple enumeration
+
+For events that have a narrow range of possible outcomes, the outcomes can simply be enumerated.
+
+#### Example: Weather tomorrow
+
+Tomorrow's weather can be represented as the set of strings `[sunny, cloudy, rainy]`.
+
+### Range
+
+When an event has numerical outcomes that cannot be easily enumerated, they can be represented as a series with:
+
+- start: the first possible outcome number
+- count: the number of possible outcomes
+- step: the increment
+- unit: the unit of the outcome value
```

Comment:
I don't see the point of including the units, this doesn't change what the oracle is signing and should be implicitly understood. For example, if you're signing BTC/USD it should be understood that you are signing for dollars

Reply:
That was a request from @nkohen and I think the point is that this ensures that the oracle cannot cheat (or that there can be no mistake) by signing something in a different unit. I think personally it doesn't hurt to have it to make things clear.

Reply:
I think there is no need for both "precision" and "unit". That is the point of a `unit` after all. If you want meters it's `m`, centimeters it's `cm` and so on. The unit is the precision.

Reply:
@benthecarman This leaves wiggle room to counter fraud proofs as they could just claim to sign a different unit. And I agree that given `unit`'s encoding below as a String, there's no reason not to include the precision in the unit (i.e. if you want a precision of 100 sats you can do `"100 Satoshis"` or `"microBitcoin"`

Reply:
@LLFourn
> I think there is no need for both "precision" and "unit". That is the point of a unit after all. If you want meters it's m, centimeters it's cm and so on. The unit is the precision.

I feel it doesn't hurt to have both personally. It's not like you cannot deal with having only unit, but it can make things easier. For example, if you want to sign dollar price with two decimal, having to specify cents or centi-dollar is not very practical IMHO. While specifying `$` or `dollars` and `-2` for precision is much easier to interpret.

I don't have that strong of an opinion about it, but it seems like it can be useful/practical.

Reply:
You're probably right -- it's a tad convoluted for financial quantities etc. It could work with what nadav suggests i.e. using "mBTC" etc. but are we going to use this for financial quantities? It seems like this will be for small ranges (otherwise you are going to be using the other one).

Maybe the right solution is to wait until someone actually wants to use this for something before specifying it?

Reply:
If we think it can be useful, I would rather keep it. Since you plan on using JSON I would say it doesn't matter since you can just not put it, but for TLVs it means if we want to add it later it will be a backward incompatible change, while not using it doesn't cost anything (apart from a few bytes).